workflow_scheduler.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"encoding/json"
+	"github.com/microsoft/BladeMonRT/workflows"
+	"github.com/microsoft/BladeMonRT/logging"
+	"log"
+)
+
+/** Class for scheduling workflows. */
+type WorkflowScheduler struct {
+	schedules []interface{}
+	logger *log.Logger
+}
+
+/** Class for the schedule description in the JSON. */
+type ScheduleDescription struct {
+    Name string `json:"name"`
+    ScheduleType string `json:"schedule_type"`
+    Workflow string `json:"workflow"`
+    Enable bool `json:"enable"`
+
+	// Attributes specific to events of type 'on_win_event'.
+	WinEventSubscribeQueries [][]string `json:"win_event_subscribe_queries"`
+}
+
+/** Class that represents a query for subscribing to a windows event. */
+type WinEventSubscribeQuery struct {
+	channel string
+	query string
+}
+
+func (workflowScheduler *WorkflowScheduler) addWinEventBasedSchedule(workflow workflows.InterfaceWorkflow, eventQueries []WinEventSubscribeQuery) {
+	workflowScheduler.logger.Println("Workflow:", workflow)
+	for _, eventQuery := range eventQueries {
+		workflowScheduler.logger.Println("Channel:", eventQuery.channel)
+		workflowScheduler.logger.Println("Query:", eventQuery.query)
+		// TO DO: Subscribe to an event using the gowinlog library
+	}
+}
+
+func newWorkflowScheduler(schedulesJson []byte, workflowFactory WorkflowFactory) WorkflowScheduler {
+	var logger *log.Logger = logging.LoggerFactory{}.ConstructLogger("WorkflowScheduler")
+	var workflowScheduler *WorkflowScheduler = &WorkflowScheduler{logger: logger}
+
+	// Parse the schedules JSON and add the schedules to the workflow scheduler.
+	var schedules map[string][]ScheduleDescription
+	json.Unmarshal([]byte(schedulesJson), &schedules)
+	for _, schedule := range schedules["schedules"] {
+		switch schedule.ScheduleType {
+			case "on_win_event":
+				var workflow workflows.InterfaceWorkflow = workflowFactory.constructWorkflow(schedule.Workflow)	
+				var eventQueries []WinEventSubscribeQuery = parseEventSubscribeQueries(schedule.WinEventSubscribeQueries)			
+				workflowScheduler.addWinEventBasedSchedule(workflow, eventQueries) 
+			default:
+				panic("Given schedule type not supported")
```

Comment:
Panics are okay here during initialization

but during workflow execution they are off limits because we don't want to restart BRT everytime a workflow is executed. Not sure if you had panics elsewhere, but double check that they only happen during initialization